docs/source/dev-guide/plugin-api/index.rst

```diff
@@ -0,0 +1,132 @@
+=======
+Plugins
+=======
+
+As of version ``4.14.0``, ``conda`` has support for user plugins, enabling extension and/or
+alterations to some of its functionality.
+
+An overview of ``pluggy``
+-------------------------
+
+Plugins in ``conda`` are implemented with the use of Pluggy_, a Python framework used by
+other projects, such as ``pytest``, ``tox``, and ``devpi``. ``pluggy`` provides the ability to
+extend and modify the behavior of ``conda`` via function hooking, which results in plugin
+systems that are discoverable with the use of `Python package entrypoints`_.
+
+At its core, creating and implementing custom plugins with the use of ``pluggy`` can
+be broken down into two parts:
+
+1. Define the hooks you want to register
+2. Register your plugin under the ``conda`` entrypoint namespace
+
+If you would like more information about ``pluggy``, please refer to their documentation_
+for a full description of its features.
+
+
+Basic hook and entry point examples
+-----------------------------------
+
+Hook
+~~~~
+
+Below is an example of a very basic plugin "hook":
+
+.. code-block:: python
+   :caption: my_plugin.py
+
+   import conda.plugins
+
+
+   @conda.plugins.register
+   def conda_subcommands():
+       ...
```

Comment:
So each hook implementation is supposed to have the same name and number of arguments as a hook spec. So far there is only a "conda_subcommands" hook spec. pluggy only checks the name and number of arguments, and doesn't do any e.g. `isinstance` or Python annotations checks.

Then pluggy makes it easy to call every registered `conda_subcommands` function by calling `pluginmanager.hook.conda_subcommands(arguments)`, (in this case no arguments), passes any arguments to all the plugins, and returns a list (in this case a list of lists) of results.

Reply:
correct, we have some ideas for making this cleaner/safer but yes that's roughly how pluggy works

Reply:
it should be straightforward to e.g. read type annotations off the collected hooks instead of calling them right away, if we wanted to do pluggy++

Reply:
maybe the stdlib doesn't have 'are these two type annotations compatible' functions

Reply:
but since this is Python we should also follow duck typing, looks like a plugin, acts like a plugin 🤷🏼‍♂️